sound/soc/sof/intel/hda-dsp.c

```diff
@@ -431,9 +503,19 @@ static int hda_dsp_set_D0_state(struct snd_sof_dev *sdev,
 		    !hda_enable_trace_D0I3_S0 ||
 		    sdev->system_suspend_target != SOF_SUSPEND_NONE)
 			flags = HDA_PM_NO_DMA_TRACE;
+
+		/* check if links should be powered off */
+		ret = hda_dsp_link_power_save_restore(sdev, target_state->substate);
+		if (ret < 0)
+			return ret;
 	} else {
 		/* prevent power gating in D0I0 */
 		flags = HDA_PM_PPG;
+
+		/* check if links should be powered back on */
+		ret = hda_dsp_link_power_save_restore(sdev, target_state->substate);
+		if (ret < 0)
+			return ret;
```

Comment:
Lines 507-510 and 516-518 are now identical and can be put outside the if-else.

Reply:
in fact I think I'd rather split that function into two, because there seems to be literally 0 shared code. It's like
```
static void fn(int x)
{
	if (x == A) {
		do_A_stuff();
		return;
	}
	do_B_stuff();
}

static caller(void)
{
	if (x == A) {
		do_some_A_stuff();
		fn(x);
	} else {
		fn(x);
	}
```
i.e. the caller already checked the condition, why check it in the called function again if there's really nothing that the called function has to do for both cases?

Reply:
ack, done!